src/main/permissionsManager.ts

```diff
@@ -0,0 +1,170 @@
+import {Logger} from 'common/log';
+import {t} from 'common/utils/util';
+import {isTrustedURL, parseURL} from 'common/utils/url';
+
+import {permissionsJson} from 'main/constants';
+import {localizeMessage} from 'main/i18nManager';
+import ViewManager from 'main/views/viewManager';
+import CallsWidgetWindow from 'main/windows/callsWidgetWindow';
+import MainWindow from 'main/windows/mainWindow';
+
+const log = new Logger('PermissionsManager');
+
+// supported permission types
+const supportedPermissionTypes = [
+    'media',
+    'geolocation',
+    'notifications',
+    'fullscreen',
+    'openExternal',
+    'clipboard-sanitized-write',
+];
+
+// permissions that require a dialog
+const authorizablePermissionTypes = [
+    'media',
+    'geolocation',
+    'notifications',
+];
+
+type Permissions = {
+    [origin: string]: {
+        [permission: string]: {
+            allowed: boolean;
+            alwaysDeny?: boolean;
+        };
+    };
+};
+
+export class PermissionsManager extends JsonFileManager<Permissions> {
+    handlePermissionRequest = async (
+        webContents: WebContents,
+        permission: string,
+        callback: (granted: boolean) => void,
+        details: PermissionRequestHandlerHandlerDetails,
+    ) => {
+        callback(await this.doPermissionRequest(
+            webContents.id,
+            permission,
+            details.securityOrigin ?? details.requestingUrl,
+        ));
+    }
+
+    doPermissionRequest = async (
+        webContentsId: number,
+        permission: string,
+        requestingURL: string,
+    ) => {
+        log.debug('doPermissionRequest', requestingURL, permission);
+
+        // is the requested permission type supported?
+        if (!supportedPermissionTypes.includes(permission)) {
+            return false;
+        }
+
+        // allow if the request is coming from the local renderer process instead of the remote one
+        const mainWindow = MainWindow.get();
+        if (mainWindow && webContentsId === mainWindow.webContents.id) {
+            return true;
+        }
+
+        const parsedURL = parseURL(requestingURL);
+        if (!parsedURL) {
+            return false;
+        }
+
+        let serverURL;
+        if (CallsWidgetWindow.isCallsWidget(webContentsId)) {
+            serverURL = CallsWidgetWindow.getViewURL();
+        } else {
+            serverURL = ViewManager.getViewByWebContentsId(webContentsId)?.view.server.url;
+        }
+
+        if (!serverURL) {
+            return false;
+        }
+
+        // is the requesting url trusted?
+        if (!isTrustedURL(parsedURL, serverURL)) {
+            return false;
+        }
+
+        // For certain permission types, we need to confirm with the user
+        if (authorizablePermissionTypes.includes(permission)) {
+            const currentPermission = this.json[parsedURL.origin]?.[permission];
+
+            // If previously allowed, just allow
+            if (currentPermission?.allowed) {
```

Comment:
Nothing to change here. I just wanted to mention that I appreciate the way these are broken up since it makes it very easy to understand the flow through this method